homeassistant/bootstrap.py

```diff
@@ -150,28 +158,32 @@ def from_config_dict(config: Dict[str, Any],
     _LOGGER.info("Home Assistant core initialized")
 
     # stage 0, load logging components
-    for component in components:
-        if component in LOGGING_COMPONENT:
-            hass.async_create_task(
-                async_setup_component(hass, component, config))
-
-    await hass.async_block_till_done()
+    await asyncio.gather(*[
+        async_setup_component(hass, component, config)
+        for component in components
+        if component in LOGGING_COMPONENT
+    ])
+    # I have audited the logging components they don't create any
+    # critical tasks so we don't need to block_till_done here.
 
     # stage 1
-    for component in components:
-        if component in FIRST_INIT_COMPONENT:
-            hass.async_create_task(
-                async_setup_component(hass, component, config))
+    asyncio.gather(*[
+        async_setup_component(hass, component, config)
+        for component in components
+        if component in FIRST_INIT_COMPONENT
+    ])
 
+    # Await any side tasks created in stage 1, plus loading the registries
```

Comment:
You're playing with fire now. This only works because async_get_registry calls create_task. Now bootstrap relies on an implementation detail, that's not good.

We should await the registries here and store them on the hass object. That way we can start using them from callback functions too.

Reply:
> This only works because async_get_registry calls create_task. 

No it doesn't (I removed the task creation bit). Gather creates and starts the tasks, then you await the awaitable that it returns to wait for them to finish. I've proved this by putting a sleep after the gather but before the awake, the registry will load before the sleep finishes.